File: 44.py
# -*- coding: utf-8 -*-
# @Author: Lich Amnesia
# @Email: alwaysxiaop@gmail.com
# @Date:   2016-10-28 16:33:58
# @Last Modified by:   Lich Amnesia
# @Last Modified time: 2016-10-28 16:34:03
# MLE

class Solution(object):

    def isMatch(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: bool
        """
        ns = len(s)
        np = len(p)
        if not ns and not np:
            return True
        dp = [[False] * (np + 1) for x in range(ns + 1)]
        # print dp
        dp[0][0] = True
        for i in range(1, np + 1):
            dp[0][i] = dp[0][i - 1] and p[i - 1] == '*'
        for i in range(1, ns + 1):
            dp[i][0] = False
        for i in range(1, ns + 1):
            for j in range(1, np + 1):
                if p[j - 1] == '?' or p[j - 1] == s[i - 1]:
                    dp[i][j] = dp[i - 1][j - 1]
                elif p[j - 1] == '*':
                    dp[i][j] = dp[i - 1][j] or dp[i][j - 1]
        return dp[ns][np]

File: java/44.java
/*
* @Author: Lich Amnesia
* @Date:   2016-10-28 16:34:07
* @Last Modified by:   Lich Amnesia
* @Last Modified time: 2016-10-28 16:42:08
*/

public class Solution {
    public boolean isMatch(String s, String p) {
        int m = s.length(), n = p.length();
        char[] ws = s.toCharArray();
        char[] wp = p.toCharArray();
        boolean[][] dp = new boolean[m+1][n+1];
        dp[0][0] = true;
        for (int i = 1; i < n + 1; i ++){
            dp[0][i] = dp[0][i - 1] && wp[i - 1] == '*';
        }
        for (int i = 1; i < m + 1; i ++){
            dp[i][0] = false;
        }
        for (int i = 1; i < m + 1; i ++){
            for (int j = 1; j < n + 1; j ++){
                if (wp[j - 1] == '?' || wp[j - 1] == ws[i - 1])
                    dp[i][j] = dp[i - 1][j - 1];
                else if (wp[j - 1] == '*')
                    dp[i][j] = dp[i - 1][j] || dp[i][j - 1];
            }
        }
        return dp[m][n];
    }
}

